### PR TITLE
Docs: Added Tabular Custom Model tutorial

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*
#1214 

*Description of changes:*

- Added custom model tutorial to Tabular docs

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1238-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1238/1/index.html

**Comment by szha** [comment]
Job PR-1238-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1238/2/index.html

File: docs/tutorials/tabular_prediction/tabular-custom-model.md [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+  - Refer to [NNFastAiTabularModel](../../_modules/autogluon/tabular/models/fastainn/tabular_nn_fastai.html#NNFastAiTabularModel)
+- Advanced problem type support (Softclass, Quantile)
+  - Refer to RandomForestModel
+- Text feature type support
+  - Refer to [TextPredictorModel](../../_modules/autogluon/tabular/models/text_prediction/text_prediction_v1_model.html#TextPredictorModel)
+- Image feature type support
+  - Refer to [ImagePredictorModel](../../_modules/autogluon/tabular/models/image_prediction/image_predictor.html#ImagePredictorModel)
+- Custom HPO logic
+  - Refer to LGBModel
+
```
Comment (gradientsky) [inline docs/tutorials/tabular_prediction/tabular-custom-model.md:34]: Can we make this a table?

aspect | reference implementation
-------|--------------------------
Respecting time limit / early stopping logic | [LGBModel](../../_modules/autogluon/tabular/models/lgb/lgb_model.html#LGBModel) and [RandomForestModel](../../_modules/autogluon/tabular/models/rf/rf_model.html#RFModel)

File: docs/tutorials/tabular_prediction/tabular-custom-model.md [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+class CustomRandomForestModel(AbstractModel):
+    def __init__(self, **kwargs):
+        # Simply pass along kwargs to parent, and init our internal `_feature_generator` variable to None
+        super().__init__(**kwargs)
+        self._feature_generator = None
+
+    # The `_preprocess` method takes the input data and transforms it to the internal representation usable by the model.
+    # `_preprocess` is called by `preprocess` and is used during model fit and model inference.
+    def _preprocess(self, X: pd.DataFrame, is_train=False, **kwargs) -> np.ndarray:
+        print(f'Entering the `_preprocess` method: {len(X)} rows of data (is_train={is_train})')
```
Comment (gradientsky) [inline docs/tutorials/tabular_prediction/tabular-custom-model.md:59]: prints are spamming logs in the docs in later examples

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 1 above)
Comment (Innixma) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md:34]: Do you know how to do that in the docs?

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 2 above)
Comment (Innixma) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md:59]: @jwmueller thoughts on this? The print statements are meant to help the user understand how and when the methods are called.

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 1 above)
Comment (jwmueller) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md:34]: Agreed the current bullet pt formatting is confusing because nested bullets appear to be part of sequential bulleted list in the docs.  Could alternatively just reformat it like this: 

Respecting time limit / early stopping logic:  Refer to LGBModel
Respecting memory usage limit:  Refer to RandomForestModel
...

File: docs/tutorials/index.rst [hunk 3]
```diff
@@ -82,6 +82,12 @@ Tabular Prediction
 
       How to predict multiple columns in a data table.
 
+   .. card::
+      :title: Adding a Custom Model
+      :link: tabular-custom-model.html
+
+      How to add a custom model to AutoGluon.
```
Comment (jwmueller) [inline docs/tutorials/index.rst:89]: custom tabular model

File: docs/tutorials/tabular_prediction/tabular-custom-model.md [hunk 4]
```diff
@@ -0,0 +1,374 @@
+# Adding a custom model to AutoGluon
+:label:`sec_tabularcustommodel`
+
+**Tip**: If you are new to AutoGluon, review :ref:`sec_tabularquick` to learn the basics of the AutoGluon API.
+
+This tutorial describes how to add a custom model to AutoGluon that can be trained and ensembled alongside the default models ([default model documentation](../../api/autogluon.tabular.models.html#module-autogluon.tabular.models)).
```
Comment (jwmueller) [inline docs/tutorials/tabular_prediction/tabular-custom-model.md:6]: trained, hyperparameter-tuned, and ensembled

File: docs/tutorials/tabular_prediction/tabular-custom-model.md [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+
+```{.python .input}
+from autogluon.core.models import BaggedEnsembleModel
+bagged_custom_model = BaggedEnsembleModel(CustomRandomForestModel())
+bagged_custom_model.fit(X=X_clean, y=y_clean, k_fold=10)  # Perform 10-fold bagging
+bagged_score = bagged_custom_model.score(X_test_clean, y_test_clean)
+print(f'test score ({bagged_custom_model.eval_metric.name}) = {bagged_score} (bagged)')
+print(f'Bagging increased model accuracy by {round(bagged_score - score, 4) * 100}%!')
+```
+
```
Comment (jwmueller) [inline docs/tutorials/tabular_prediction/tabular-custom-model.md:267]: might need to clarify how the data preprocessing works for bagged models. I recall there was a bug in one model at some point because it assumed the preprocessing was being re-run for each bagging fold.

File: docs/tutorials/tabular_prediction/tabular-custom-model.md [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+```{.python .input}
+from autogluon.features.generators import AutoMLPipelineFeatureGenerator
+
+feature_generator = AutoMLPipelineFeatureGenerator()
+X_clean = feature_generator.fit_transform(X)
+
+X_clean.head(5)
+```
+
```
Comment (jwmueller) [inline docs/tutorials/tabular_prediction/tabular-custom-model.md:198]: may be helpful to clarify what AutoMLPipelineFeatureGenerator does not do for you, but many models might like to have (eg. rescaling/imputation of numerical features)

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 2 above)
Comment (jwmueller) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md:59]: I think these statements are very helpful. But if you really don't want them appear in later logs, you could give the model a hyperparameter that controls whether these are printed or not

File: docs/tutorials/tabular_prediction/tabular-custom-model.md [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+This tutorial describes how to add a custom model to AutoGluon that can be trained and ensembled alongside the default models ([default model documentation](../../api/autogluon.tabular.models.html#module-autogluon.tabular.models)).
+
+In this example, we create a custom Random Forest model for use in AutoGluon. All models in AutoGluon inherit from the AbstractModel class ([AbstractModel source code](../../_modules/autogluon/core/models/abstract/abstract_model.html)), and must follow its API to work alongside other models.
+
+Note that while this tutorial provides a basic model implementation, this does not cover many aspects that are used in most implemented models.
+
+To best understand how to implement more advanced functionality, refer to the [source code](../../api/autogluon.tabular.models.html#module-autogluon.tabular.models) of the following models:
+
+- Respecting time limit / early stopping logic
```
Comment (jwmueller) [inline docs/tutorials/tabular_prediction/tabular-custom-model.md:14]: may add to this list: Example of lazy import of package dependencies

File: docs/tutorials/tabular_prediction/tabular-custom-model.md [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+custom_model.fit(X=X_clean, y=y_clean)  # Fit custom model
+
+# To save to disk and load the model, do the following:
+# load_path = custom_model.path
+# custom_model.save()
+# del custom_model
+# custom_model = CustomRandomForestModel.load(path=load_path)
+```
+
+### Predict with fit model
```
Comment (jwmueller) [inline docs/tutorials/tabular_prediction/tabular-custom-model.md:216]: when it's being used in past-tense, I'd replace "fit" with "previously fit" or "trained", otherwise can be confusing

File: docs/tutorials/tabular_prediction/tabular-custom-model.md [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+    # The `_fit` method takes the input training data (and optionally the validation data) and trains the model.
+    def _fit(self,
+             X: pd.DataFrame,  # training data
+             y: pd.Series,  # training labels
+             # X_val=None,  # val data (unused in RF model)
+             # y_val=None,  # val labels (unused in RF model)
+             # time_limit=None,  # time limit in seconds (ignored in tutorial)
+             **kwargs):  # kwargs includes many other potential inputs, refer to AbstractModel documentation for details
+        print('Entering the `_fit` method')
+        from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
```
Comment (jwmueller) [inline docs/tutorials/tabular_prediction/tabular-custom-model.md:85]: demonstrating lazy import here might accelerate contributions

File: docs/tutorials/tabular_prediction/tabular-indepth.md [hunk 10]
```diff
@@ -1,7 +1,7 @@
 # Predicting Columns in a Table - In Depth
 :label:`sec_tabularadvanced`
 
-**Tip**: If you are new to AutoGluon, review :ref:`sec_tabularquick` to learn the basics of the AutoGluon API.
+**Tip**: If you are new to AutoGluon, review :ref:`sec_tabularquick` to learn the basics of the AutoGluon API. To learn how to implement custom models, review :ref:`sec_tabularcustommodel`.
```
Comment (jwmueller) [inline docs/tutorials/tabular_prediction/tabular-indepth.md:4]: "implement custom models" may be clearer as:
"add your own custom models to the set that AutoGluon trains, tunes, and ensembles"

**Review by jwmueller — APPROVED** [review]
just have minor language/formatting edits, otherwise looks great. Hopefully this'll make external contributions way easier!

File: docs/tutorials/tabular_prediction/tabular-indepth.md (hunk 10 above)
Comment (jwmueller) [reply docs/tutorials/tabular_prediction/tabular-indepth.md:4]: Maybe also add a FAQ question with this link as the answer such as:

"Does AutoGluon support my favorite model XYZ"?

File: docs/tutorials/index.rst (hunk 3 above)
Comment (Innixma) [reply docs/tutorials/index.rst:89]: With the shift of AbstractLearner and AbstractTrainer to core, I think this will be more relevant to adding custom models  in general, not just tabular.

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 4 above)
Comment (Innixma) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md:6]: added

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 5 above)
Comment (Innixma) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md:267]: Added a bit of clarification on what happens inside the bagged model, but I don't think we need to mention the data preprocessing logic, as we currently have users work with `_preprocess` which works on a per-fold basis and is much safer.

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 6 above)
Comment (Innixma) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md:198]: added

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 7 above)
Comment (Innixma) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md:14]: added

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 8 above)
Comment (Innixma) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md:216]: replaced with 'trained'

File: docs/tutorials/tabular_prediction/tabular-indepth.md (hunk 10 above)
Comment (Innixma) [reply docs/tutorials/tabular_prediction/tabular-indepth.md:4]: updated and added to FAQ

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 9 above)
Comment (Innixma) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md:85]: added extra code comments for this

**Comment by szha** [comment]
Job PR-1238-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1238/3/index.html

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 1 above)
Comment (Innixma) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md]: Something is weird with D2L's formatting... I added a table but it also has issues with spacing that it shouldn't. Similarly, the nested bullets look fine in my PyCharm preview of markdown, but don't work in our built docs.

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 2 above)
Comment (Innixma) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md:54]: I'll plan to keep them for now, adding it as a hyperparameter would maybe detract from tutorial, but we can revisit later if we think controlling it is cleaner.

File: docs/tutorials/tabular_prediction/tabular-custom-model.md (hunk 1 above)
Comment (Innixma) [reply docs/tutorials/tabular_prediction/tabular-custom-model.md]: I'll merge for now, but we should really find a better way to do these things... I don't understand why what we are using has less working functionality than default markdown.
